modeling/algorithm/algorithms_test.go

```diff
@@ -0,0 +1,110 @@
+)
+
+type testCase struct {
+	name           string
+	input          []types.Zone
+	expectedOutput map[string]types.EndpointSliceGroup
+	expectedErr    error
+}
+type routingAlgorithmTest struct {
+	algName   string
+	alg       RoutingAlgorithm
+	testCases []testCase
+}
+
+func (algTest routingAlgorithmTest) doTest(t *testing.T) {
+	for _, testcase := range algTest.testCases {
+		t.Run(testcase.name, func(t *testing.T) {
+			region, err := types.CreateRegionInfo(testcase.input)
+			if err != nil {
+				t.Errorf("[test %s] encountered unexpected error while creating RegionInfo with %+v", algTest.algName, testcase.input)
+				return
+			}
+			sliceGroups, err := algTest.alg.CreateSliceGroups(region)
+			if !reflect.DeepEqual(err, testcase.expectedErr) {
+				t.Errorf("[test %s] got error: %v, expected err: %v", algTest.algName, err, testcase.expectedErr)
+				return
+			}
+			if !compareSliceGroups(sliceGroups, testcase.expectedOutput, false) {
+				t.Errorf("[test %s] got slices: %+v, expected slices: %+v", algTest.algName, sliceGroups, testcase.expectedOutput)
+				return
+			}
+		})
+	}
+}
+
+// helper function to compare float64 that if two floats are within epsilon
+// delta, we deem them as equal
+func compareFloat(a float64, b float64, epsilon float64) bool {
+	return math.Abs(a-b) <= epsilon
+}
+
+// helper function to compare EndpointSliceGroups, it takes zero entries in the
+// map into consideration, i.e. "key" : zeroValue will be considered equal to
+// "key" doesn't exist
+func compareSliceGroups(sliceGroupsA map[string]types.EndpointSliceGroup, sliceGroupsB map[string]types.EndpointSliceGroup, reverse bool) bool {
+	if (sliceGroupsA == nil && sliceGroupsB != nil) || (sliceGroupsA != nil && sliceGroupsB == nil) {
+		return false
+	}
+	// deep comparison on every key/value pair, we take missing key and key:
+	// zoneValue as equal.
+	for key, sliceGroupB := range sliceGroupsB {
+		sliceGroupA := sliceGroupsA[key]
+		if sliceGroupA.Label != sliceGroupB.Label {
+			return false
+		}
+		if (sliceGroupA.Composition == nil && sliceGroupB.Composition != nil) || (sliceGroupA.Composition != nil && sliceGroupB.Composition == nil) {
+			return false
+		}
+		for zone, compB := range sliceGroupB.Composition {
+			compA := sliceGroupA.Composition[zone]
+			if compA.Number != compB.Number {
+				return false
+			}
+			// if number = 0, weight doesn't matter
+			if !compareFloat(compA.Weight, compB.Weight, 0.00001) {
+				return (compA.Number == 0 && compB.Number == 0) || false
+			}
+		}
+		if (sliceGroupA.ZoneTrafficWeights == nil && sliceGroupB.ZoneTrafficWeights != nil) || (sliceGroupA.ZoneTrafficWeights != nil && sliceGroupB.ZoneTrafficWeights == nil) {
+			return false
+		}
+		for zone, weightB := range sliceGroupB.ZoneTrafficWeights {
+			weightA := sliceGroupA.ZoneTrafficWeights[zone]
+			if !compareFloat(weightA, weightB, 0.00001) {
+				return false
+			}
+		}
+	}
+	if reverse {
+		return true
+	}
```

Comment:
When would you only want to do a one way comparison?

Reply:
refactored this function